From 480b61538f4cad88abe28f3569b803d53de32bc5 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Wed, 22 May 2024 18:39:20 +0200
Subject: [PATCH 0975/1695] drivers: stm32_cpu_opp: skip OPP with unsupported
 frequency

Skip the OPP, if the associated frequency is not supported.

This patch allows to simplified the SoC and board device tree, as the
the list of OPP will be always present at SoC level but the list of
supported frequency for CPU are listed in RCC nodes "st,ck_mpu" and
sub-node of "st,clk_opp".

This behavior solves issue when a new OPP is added in SoC level
without modification on existing board.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 0d08c4b25171a064e87dda35cced472424c8d0c1 drivers: stm32_cpu_opp: skip OPP with unsupported frequency]
Change-Id: I6f9ac3f6c37705c4a9cd9c84552364439d9807e4
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/382034
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_cpu_opp.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/core/drivers/stm32_cpu_opp.c b/core/drivers/stm32_cpu_opp.c
index d10170439..976efbaaf 100644
--- a/core/drivers/stm32_cpu_opp.c
+++ b/core/drivers/stm32_cpu_opp.c
@@ -570,6 +570,7 @@ static TEE_Result stm32_cpu_opp_get_dt_subnode(const void *fdt, int node)
 {
 	const fdt64_t *cuint64 = NULL;
 	const fdt32_t *cuint32 = NULL;
+	uint64_t freq_hz = 0;
 	uint64_t freq_khz = 0;
 	uint64_t freq_khz_opp_def = 0;
 	uint32_t volt_uv = 0;
@@ -598,8 +599,8 @@ static TEE_Result stm32_cpu_opp_get_dt_subnode(const void *fdt, int node)
 			EMSG("Missing opp-hz");
 			return TEE_ERROR_GENERIC;
 		}
-
-		freq_khz = fdt64_to_cpu(*cuint64) / 1000ULL;
+		freq_hz = fdt64_to_cpu(*cuint64);
+		freq_khz = freq_hz / 1000ULL;
 		if (freq_khz > (uint64_t)UINT32_MAX) {
 			EMSG("Invalid opp-hz %"PRIu64, freq_khz);
 			return TEE_ERROR_GENERIC;
@@ -613,6 +614,14 @@ static TEE_Result stm32_cpu_opp_get_dt_subnode(const void *fdt, int node)
 
 		volt_uv = fdt32_to_cpu(*cuint32);
 
+		/* skip OPP when frequency is not supported */
+		if (freq_hz != clk_round_rate(cpu_opp.clock, freq_hz)) {
+			DMSG("Skip OPP %"PRIu64"kHz/%"PRIu32"uV",
+			     freq_khz, volt_uv);
+			cpu_opp.opp_count--;
+			continue;
+		}
+
 		/* skip OPP when voltage is not supported */
 		if (!opp_voltage_is_supported(cpu_opp.regul, volt_uv)) {
 			DMSG("Skip OPP %"PRIu64"kHz/%"PRIu32"uV",
-- 
2.47.2

